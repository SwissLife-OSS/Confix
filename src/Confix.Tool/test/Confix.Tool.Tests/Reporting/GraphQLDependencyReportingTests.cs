using Confix.Inputs;
using Confix.Tool.Commands.Temp;
using Confix.Tool.Schema;
using Confix.Utilities;
using Confix.Utilities.FileSystem;
using Moq;

namespace Confix.Reporting;

public class GraphQLDependencyReportingTests
{
    private readonly Mock<IGitService> _git;
    private readonly TestConfixCommandline _cli;

    public GraphQLDependencyReportingTests()
    {
        _git = new Mock<IGitService>();
        _cli = new TestConfixCommandline(x => x.AddTestService<IGitService>(_ => _git.Object));
    }

    [Fact]
    public async Task Should_OutputDependency_When_Match()
    {
        // arrange
        _cli.Directories.Home.CreateConfixRc(_confixRc);
        _cli.Directories.Content
            .CreateConfixProject(path: $"src/project/{FileNames.ConfixProject}");
        _cli.Directories.Content.CreateFileInPath("src/project/test.json",
            """
            {
                "test": {
                    "value": "The date is 2023-10-19"
                }
            }
            """
        );
        CreateComponent(_cli,
            """
                type Query {
                    value: String @dependency(kind: "foogaloo")
                }
            """);
        _cli.Directories.Content.CreateConfixSolution();
        _cli.ExecutionContext = _cli.ExecutionContext with
        {
            CurrentDirectory = _cli.Directories.Content
                .Append("src")
                .Append("project")
        };

        _git.SetupGitRoot(_cli.Directories.Content);
        _git.SetupGitBranch();
        _git.SetupGitTags();
        _git.SetupGitRepoInfo();
        _git.SetupOriginUrl();

        await _cli.RunAsync("build");
        _cli.ResetConsole();
        // act
        await _cli.RunAsync("project report");

        // assert
        SnapshotBuilder.New()
            .AddOutput(_cli)
            .RemoveDateTimes()
            .MatchSnapshot();
    }

    [Fact]
    public async Task Should_OutputDependency_When_MultipleMatch()
    {
        // arrange
        _cli.Directories.Home.CreateConfixRc(_confixRc);
        _cli.Directories.Content
            .CreateConfixProject(path: $"src/project/{FileNames.ConfixProject}");
        _cli.Directories.Content.CreateFileInPath("src/project/test.json",
            """
            {
                "test": {
                    "value": "The date is 2023-10-19"
                }
            }
            """
        );
        CreateComponent(_cli,
            """
                type Query {
                    value: String @dependency(kind: "frist")
                        @dependency(kind: "another")
                }
            """);
        _cli.Directories.Content.CreateConfixSolution();
        _cli.ExecutionContext = _cli.ExecutionContext with
        {
            CurrentDirectory = _cli.Directories.Content
                .Append("src")
                .Append("project")
        };

        _git.SetupGitRoot(_cli.Directories.Content);
        _git.SetupGitBranch();
        _git.SetupGitTags();
        _git.SetupGitRepoInfo();
        _git.SetupOriginUrl();

        await _cli.RunAsync("build");
        _cli.ResetConsole();
        // act
        await _cli.RunAsync("project report");

        // assert
        SnapshotBuilder.New()
            .AddOutput(_cli)
            .RemoveDateTimes()
            .MatchSnapshot();
    }

    [Fact]
    public async Task Should_NotOutputDependency_When_NoMatch()
    {
        // arrange
        _cli.Directories.Home.CreateConfixRc(_confixRc);
        _cli.Directories.Content
            .CreateConfixProject(path: $"src/project/{FileNames.ConfixProject}");
        _cli.Directories.Content.CreateFileInPath("src/project/test.json",
            """
            {
                "test": {
                    "value": "The date is 2023-10-19"
                }
            }
            """
        );
        CreateComponent(_cli,
            """
                type Query {
                    bar: String @dependency(kind: "foogaloo")
                }
            """);
        _cli.Directories.Content.CreateConfixSolution();
        _cli.ExecutionContext = _cli.ExecutionContext with
        {
            CurrentDirectory = _cli.Directories.Content
                .Append("src")
                .Append("project")
        };

        _git.SetupGitRoot(_cli.Directories.Content);
        _git.SetupGitBranch();
        _git.SetupGitTags();
        _git.SetupGitRepoInfo();
        _git.SetupOriginUrl();

        await _cli.RunAsync("build");
        _cli.ResetConsole();
        // act
        await _cli.RunAsync("project report");

        // assert
        SnapshotBuilder.New()
            .AddOutput(_cli)
            .RemoveDateTimes()
            .MatchSnapshot();
    }

    private static void CreateComponent(
        TestConfixCommandline cli,
        string graphql)
    {
        cli.Directories.Content.CreateConfixComponent("test", "src/project");
        var componentDirectory = cli.Directories.Content
            .Append("src")
            .Append("project")
            .Append(FolderNames.Confix)
            .Append(FolderNames.Components)
            .Append("test");

        componentDirectory.CreateFileInPath("schema.graphql", graphql);
    }

    private const string _confixRc = """
            {
                "isRoot": true,
                "component":
                {
                     "inputs": [
                          {
                            "type": "graphql"
                          }
                        ]
                },
                "project": {
                     "componentProviders": [],
                     "configurationFiles": ["$project:/test.json"]
                },
                "reporting": {
                    "dependencies": {
                        "providers": [
                            {
                                "type": "graphql"
                            }
                        ]
                    }
                }
            }
        """;
}

file static class Extensions
{
    public static void SetupGitRoot(this Mock<IGitService> service, FileSystemInfo root)
    {
        service.Setup(
                x => x.GetRootAsync(It.IsAny<GitGetRootConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => root.FullName);
    }

    public static void SetupGitBranch(
        this Mock<IGitService> service,
        string branch = "master")
    {
        service.Setup(
                x => x.GetBranchAsync(It.IsAny<GitGetBranchConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => branch);
    }

    public static void SetupGitTags(
        this Mock<IGitService> service,
        string[]? tags = default!)
    {
        tags ??= new[]
        {
            "1.0.0",
            "1.1.0"
        };
        service.Setup(
                x => x.GetTagsAsync(It.IsAny<GitGetTagsConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => tags);
    }

    public static void SetupGitRepoInfo(
        this Mock<IGitService> service,
        string hash = "1234567890",
        string message = "Initial commit",
        string author = "John Doe",
        string email = "jd@gmail.com")
    {
        service.Setup(
                x => x.GetRepoInfoAsync(It.IsAny<GitGetRepoInfoConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => new GitGetRepoInfoResult(hash, message, author, email));
    }

    public static void SetupOriginUrl(
        this Mock<IGitService> service,
        string url = "https://foobar.com")
        => service.Setup(
                x => x.GetOriginUrlAsync(
                    It.IsAny<GitGetOriginUrlConfiguration>(),
                    It.IsAny<CancellationToken>()))
            .ReturnsAsync(() => url);
}
