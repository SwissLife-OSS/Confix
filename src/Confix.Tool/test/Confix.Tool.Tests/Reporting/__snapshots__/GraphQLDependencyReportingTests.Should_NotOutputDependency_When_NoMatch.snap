--------------------------------------------------
### CLI Output
--------------------------------------------------
  Configuration files of type .confixrc located at <</tmp>>/home/.confixrc
  Configuration files of type .confix.solution located at <</tmp>>/content/.confix.solution
  Configuration files of type .confix.solution located at <</tmp>>/content/src/project/.confix.project
  Replaced magic path $project:/test.json with <</tmp>>/content/src/project/test.json.
  Merged .confix.solution from .confixrc
  Merged .confix.project from .confix.solution
  Merged .confix.component from .confix.solution
✓ Running in scope Project
✓ Configuration loaded
  Found a inline configuration file '<</tmp>>/content/src/project/test.json'
✓ Active Environment is prod
  Component detected:test <</tmp>>/content/src/project/confix/components/test/.confix.component
  Configuration files of type .confixrc located at <</tmp>>/home/.confixrc
  Configuration files of type .confix.solution located at <</tmp>>/content/.confix.solution
  Configuration files of type .confix.solution located at <</tmp>>/content/src/project/.confix.project
  Configuration files of type .confix.component located at <</tmp>>/content/src/project/confix/components/test/.confix.component
  Replaced magic path $project:/test.json with <</tmp>>/content/src/project/test.json.
  Merged .confix.solution from .confixrc
  Merged .confix.project from .confix.solution
  Merged .confix.component from .confix.solution
✓ Running in scope Component
✓ Configuration loaded
  Component input 'graphql' loaded
✓ Component inputs loaded
  Building component...
  Searching in '<</tmp>>/content/src/project/confix/components/test/schema.graphql' for GraphQL Schema
   -> <</tmp>>/content/src/project/confix/components/test/schema.graphql
  GraphQL Schema was found: '<</tmp>>/content/src/project/confix/components/test/schema.graphql' <</tmp>>/content/src/project/confix/components/test/schema.graphql
  Replacing existing schema.json file: '<</tmp>>/content/src/project/confix/components/test/schema.json' <</tmp>>/content/src/project/confix/components/test/schema.json
  Generated schema based on GraphQL Schema:'<</tmp>>/content/src/project/confix/components/test/schema.json' <</tmp>>/content/src/project/confix/components/test/schema.json
  Building component completed
✓ Component inputs loaded
  Dependency provider 'graphql' loaded
✓ Dependency providers loaded
  Commit: 1234567890 by John Doe on master Initial commit
  Repository: <</tmp>> <</tmp>>/content
  Solution: content .
  Project: src.project src/project
  Found component test in provider __LOCAL
  Start loading components from project 'src.project'
✓ Loaded 1 components
  -  test
  Detected 0 variables
  Report: <</tmp>>/content/src/project/test.json prod
[
   {
      "configurationPath": "<</tmp>>/content/src/project/test.json",
      "environment": "prod",
      "timestamp": <<date>>,
      "project": {
         "name": "src.project",
         "path": "src/project"
      },
      "solution": {
         "name": "content",
         "path": "."
      },
      "repository": {
         "name": "<</tmp>>",
         "originUrl": "https://foobar.com"
      },
      "commit": {
         "hash": "1234567890",
         "message": "Initial commit",
         "author": "John Doe",
         "email": "jd@gmail.com",
         "branch": "master",
         "tags": [
            "1.0.0",
            "1.1.0"
         ]
      },
      "variables": [
      ],
      "components": [
         {
            "providerName": "__LOCAL",
            "name": "test",
            "version": "latest",
            "mountingPoints": [
               "test"
            ]
         }
      ],
      "dependencies": [
      ]
   }
]
